### PR TITLE
feat: add discord-utilities package

### PR DESCRIPTION
**Pull request by Favna**

### CODE REVIEW

File: packages/discord-utilities/rollup.config.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// import commonjs from '@rollup/plugin-commonjs';
```
Comment (vladfrangu) [inline packages/discord-utilities/rollup.config.ts:1]: ?

File: packages/discord-utilities/src/lib/regexes.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ * @remark Capture group 1 is the ID of the channel. It is named `id`.
+ */
+export const ChannelMentionRegex = /^(?:<#)?(?<id>\d{17,19})>?$/;
+
+/**
+ * Regex that matches links on the known Discord host names
+ * @raw `/^(?:\w+\.)?(?:discordapp.com|discord.gg|discord.com)$/i`
+ * @remark The regex is case insensitive
+ */
+export const DiscordHostnameRegex = /^(?:\w+\.)?(?:discordapp\.com|discord\.gg|discord\.com)$/i;
```
Comment (vladfrangu) [inline packages/discord-utilities/src/lib/regexes.ts:13]: This should:
1. Include ptb, canary
2. be optimized to not repeat `.com`, if applicable
3. Include named capture groups?
4. Why are you using `$` 🤔

File: packages/discord-utilities/src/lib/regexes.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+export const HttpUrlRegex = /^https?:\/\//;
+
+/**
+ * Regex that can capture the Guild, Channel, and Message ID based on a shareable Discord message link.
+ * @raw `/^(?:https:\/\/)?(?:ptb\.|canary\.)?discord(?:app)?\.com\/channels\/(?<guildId>\d{17,19})\/(?<channelId>\d{17,19})\/(?<messageId>\d{17,19})$/`
+ * @remark Capture group 1 is the ID of the guild the message was sent in. It is named `guildId`.
+ * @remark Capture group 2 is the ID of the channel in that guild the message was sent in. It is named `channelId`.
+ * @remark Capture group 3 is the ID of the message itself. It is named `messageId`.
+ */
+export const MessageLinkRegex = /^(?:https:\/\/)?(?:ptb\.|canary\.)?discord(?:app)?\.com\/channels\/(?<guildId>\d{17,19})\/(?<channelId>\d{17,19})\/(?<messageId>\d{17,19})$/;
```
Comment (vladfrangu) [inline packages/discord-utilities/src/lib/regexes.ts:60]: First ID can also be `@me`... should add that to the regex

File: packages/discord-utilities/src/lib/regexes.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ * @remark Capture group 1 is the Snowflake. It is named `snowflake`.
+ */
+export const SnowflakeRegex = /^(?<snowflake>\d{17,19})$/;
+
+/**
+ * Regex that can capture a Twemoji (Twitter Emoji)
+ * @raw `/^(?<name>[^a-zA-Z0-9]{1,11})$/`
+ * @remark Capture group 1 is the Twemoji name. It is named `name`.
+ */
+export const TwemojiRegex = /^(?<name>[^a-zA-Z0-9]{1,11})$/;
```
Comment (vladfrangu) [inline packages/discord-utilities/src/lib/regexes.ts:100]: this ain't it chief, if you want the regex, read twemoji's source

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/discord-utilities/src/lib/regexes.ts (hunk 4 above)
Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts:100]: Blame @kyranet for having this in Skyra then smh

File: packages/discord-utilities/src/lib/regexes.ts (hunk 2 above)
Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts:13]: > Include named capture groups?

They are non-capture groups.

Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts]: Updated

File: packages/discord-utilities/src/lib/regexes.ts (hunk 3 above)
Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts]: Fixed

File: packages/discord-utilities/rollup.config.ts (hunk 1 above)
Comment (Favna) [reply packages/discord-utilities/rollup.config.ts]: Woops

File: packages/discord-utilities/src/lib/regexes.ts (hunk 4 above)
Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts:100]: fixed

File: packages/discord-utilities/src/lib/regexes.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ * @remark Capture group 1 is the ID of the role. It is named `id`.
+ */
+export const RoleMentionRegex = /^(?:<@&)?(?<id>\d{17,19})>?$/;
+
+/**
+ * Regex that can capture any Discord Snowflake ID
+ * @raw `/^(?<snowflake>\d{17,19})$/`
+ * @remark Capture group 1 is the Snowflake. It is named `snowflake`.
+ */
+export const SnowflakeRegex = /^(?<snowflake>\d{17,19})$/;
```
Comment (kyranet) [inline packages/discord-utilities/src/lib/regexes.ts:98]:
```suggestion
 * @remark Capture group 1 is the Snowflake. It is named `id`.
 */
export const SnowflakeRegex = /^(?<id>\d{17,19})$/;
```

It'd be useful to be able to do:

```javascript
const id = (RoleMentionRegex.exec(str) ?? SnowflakeRegex.exec(str))?.groups.id ?? null;
```

To test a role mention and a ID.

File: packages/discord-utilities/src/lib/regexes.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ * @remark Capture group 2 is the ID of the emoji. It is named `id`.
+ */
+export const ParsedCustomEmojiWithGroups = /^(?<animated>a?):(?<name>[^:]+):(?<id>\d{17,19})$/;
+
+/**
+ * Regex that can capture the ID in Discord Role mentions
+ * @raw `/^(?:<@&)?(?<id>\d{17,19})>?$/`
+ * @remark Capture group 1 is the ID of the role. It is named `id`.
+ */
+export const RoleMentionRegex = /^(?:<@&)?(?<id>\d{17,19})>?$/;
```
Comment (kyranet) [inline packages/discord-utilities/src/lib/regexes.ts:91]: Those RegExps should stay pure and not allow ill-formed syntax (`<@790892715248320512` and `790892715248320512>` are invalid mentions, but this RegExp matches both of them).

```suggestion
export const RoleMentionRegex = /^<@&(?<id>\d{17,19})>$/;
```

Same goes for `ChannelMentionRegex` and `UserOrMemberMentionRegex`.

File: packages/discord-utilities/src/lib/regexes.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+ * @remark Capture group 3 is the Top-Level Domain *without* `.`. It is named `tld`.
+ */
+export const DiscordHostnameRegex = /(?<prefix>\w+\.)?(?<hostname>dis(?:cord)?(?:app|merch|status)?)\.(?<tld>com|g(?:d|g|ift)|(?:de(?:sign|v))|media|new|store|net)/i;
+
+/**
+ * Regex that can can capture the code of Discord invite links
+ * @raw `^(?:https?:\/\/)?(?:www.)?(?:discord\.gg\/|discordapp\.com\/invite\/)?(?<code>[\w\d-]{2,})$`
+ * @remark Capture group 1 is the invite URL's unique code. It is named `code`.
+ */
+export const DiscordInviteLinkRegex = /^(?:https?:\/\/)?(?:www\.)?(?:discord\.gg\/|discordapp\.com\/invite\/)?(?<code>[\w\d-]{2,})$/i;
```
Comment (kyranet) [inline packages/discord-utilities/src/lib/regexes.ts:25]: Support both discord.com and discordapp.com:

```suggestion
export const DiscordInviteLinkRegex = /^(?:https?:\/\/)?(?:www\.)?(?:discord\.gg\/|discord(?:app)?\.com\/invite\/)?(?<code>[\w\d-]{2,})$/i;
```

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/discord-utilities/src/lib/regexes.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+/**
+ * Regex that matches links on the known Discord host names
+ * @raw `/(?<prefix>\w+\.)?(?<hostname>dis(?:cord)?(?:app|merch|status)?)\.(?<tld>com|g(?:d|g|ift)|(?:de(?:sign|v))|media|new|store|net)/i`
+ * @remark The regex is case insensitive
+ * @remark Capture group 1 is whatever is in the URL *before* the Discord hostname, such as `https://` and subdomains. It is named `prefix`.
+ * @remark Capture group 2 is the hostname for this URL, primarily `discord` but can also be `discordmerch`, `discordstatus`, `dis`, and `discordapp`. It is named `hostname`.
+ * @remark Capture group 3 is the Top-Level Domain *without* `.`. It is named `tld`.
+ */
+export const DiscordHostnameRegex = /(?<prefix>\w+\.)?(?<hostname>dis(?:cord)?(?:app|merch|status)?)\.(?<tld>com|g(?:d|g|ift)|(?:de(?:sign|v))|media|new|store|net)/i;
```
Comment (vladfrangu) [inline packages/discord-utilities/src/lib/regexes.ts:18]:
```suggestion
export const DiscordHostnameRegex = /(?<subdomain>\w+\.)?(?<hostname>dis(?:cord)?(?:app|merch|status)?)\.(?<tld>com|g(?:d|g|ift)|(?:de(?:sign|v))|media|new|store|net)/i;
```

File: packages/discord-utilities/src/lib/regexes.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+ * @remark Capture group 1 is the invite URL's unique code. It is named `code`.
+ */
+export const DiscordInviteLinkRegex = /^(?:https?:\/\/)?(?:www\.)?(?:discord\.gg\/|discord(?:app)?\.com\/invite\/)?(?<code>[\w\d-]{2,})$/i;
+
+/**
+ * Regex that can capture the ID of any animated or non-animated custom Discord emoji
+ * @raw `/^(?:<a?:\w{2,32}:)?(\d{17,21})>?$/`
+ * @remark Capture group 1 is the ID of the emoji. It is named `id`.
+ */
+export const EmojiRegex = /^(?:<a?:\w{2,32}:)?(?<id>\d{17,21})>?$/;
```
Comment (vladfrangu) [inline packages/discord-utilities/src/lib/regexes.ts:32]: Maybe make a group for animated/name?

File: packages/discord-utilities/src/lib/regexes.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+ * @remark Capture group 1 is the ID of the user. It is named `id`.
+ */
+export const UserOrMemberMentionRegex = /^<@!?(?<id>\d{17,19})>?$/;
+
+/**
+ * Regex that matches any WebSocket URL starting with `ws` or `wss`
+ * @raw `/^wss?:\/\//`
+ * @remark for regular HTTP URLs see [[HttpUrlRegex]]
+ */
+export const WebsocketUrlRegex = /^wss?:\/\//;
```
Comment (vladfrangu) [inline packages/discord-utilities/src/lib/regexes.ts:118]:
```suggestion
export const WebSocketUrlRegex = /^wss?:\/\//;
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/discord-utilities/src/lib/regexes.ts (hunk 9 above)
Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts]: done in 012b53881840183a57eb0785d40e7503bf617aee

File: packages/discord-utilities/src/lib/regexes.ts (hunk 10 above)
Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts]: done in 012b53881840183a57eb0785d40e7503bf617aee

File: packages/discord-utilities/src/lib/regexes.ts (hunk 8 above)
Comment (Favna) [reply packages/discord-utilities/src/lib/regexes.ts]: done in 012b53881840183a57eb0785d40e7503bf617aee

File: packages/discord-utilities/src/lib/regexes.ts [hunk 11]
```diff
@@ -41,12 +43,12 @@ export const FormattedCustomEmoji = /<a?:\w{2,32}:\d{17,18}>/;
...
  * Regex that can capture any animated or non-animated custom Discord emoji.
  * Similar to [[FormattedCustomEmoji]] and unlike [[EmojiRegex]] can also be a substring of a larger string.
- * @raw `/^(a?):([^:]+):(\d{17,19})$/`
+ * @raw `/^(?<animated>a?):(?<name>[^:]+):(?<id>\d{17,19})$/`
  * @remark Capture group 1 can be used to determine whether the emoji is animated or not. It is named `animated`.
  * @remark Capture group 2 is the name of the emoji as it is typed in a message. It is named `name`.
  * @remark Capture group 2 is the ID of the emoji. It is named `id`.
  */
-export const FormattedCustomEmojiWithGroups = /^(?<animated>a?):(?<name>[^:]+):(?<id>\d{17,19})$/;
+export const FormattedCustomEmojiWithGroups = /^(?<animated>a?):(?<name>[^:]+):(?<id>\d{17,21})$/;
```
Comment (kyranet) [inline packages/discord-utilities/src/lib/regexes.ts:51]:
```suggestion
export const FormattedCustomEmojiWithGroups = /^(?<animated>a?):(?<name>[^:]+):(?<id>\d{17,19})$/;
```

**Review by kyranet — CHANGES_REQUESTED** [review]
Just one last thing:

**Review by kyranet — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**